src/Env.hs

```diff
@@ -0,0 +1,93 @@
+module Env where
+
+import Data.List (foldl')
+import qualified Data.Map as Map
+import Obj
+import Types
+
+-- | Add an XObj to a specific environment. TODO: rename to envInsert
+extendEnv :: Env -> String -> XObj -> Env
+extendEnv env name xobj = envAddBinding env name (Binder emptyMeta xobj)
+
+-- | Add a Binder to an environment at a specific path location.
+envInsertAt :: Env -> SymPath -> Binder -> Env
+envInsertAt env (SymPath [] name) binder =
+  envAddBinding env name binder
+envInsertAt env (SymPath (p : ps) name) xobj =
+  case Map.lookup p (envBindings env) of
+    Just (Binder meta (XObj (Mod innerEnv) i t)) ->
+      let newInnerEnv = Binder meta (XObj (Mod (envInsertAt innerEnv (SymPath ps name) xobj)) i t)
+       in env {envBindings = Map.insert p newInnerEnv (envBindings env)}
+    Just _ -> error ("Can't insert into non-module: " ++ p)
+    Nothing -> error ("Can't insert into non-existing module: " ++ p)
+
+envReplaceEnvAt :: Env -> [String] -> Env -> Env
+envReplaceEnvAt _ [] replacement = replacement
+envReplaceEnvAt env (p : ps) replacement =
+  case Map.lookup p (envBindings env) of
+    Just (Binder _ (XObj (Mod innerEnv) i t)) ->
+      let newInnerEnv = Binder emptyMeta (XObj (Mod (envReplaceEnvAt innerEnv ps replacement)) i t)
+       in env {envBindings = Map.insert p newInnerEnv (envBindings env)}
+    Just _ -> error ("Can't replace non-module: " ++ p)
+    Nothing -> error ("Can't replace non-existing module: " ++ p)
+
+-- | Add a Binder to a specific environment.
+envAddBinding :: Env -> String -> Binder -> Env
+envAddBinding env name binder = env {envBindings = Map.insert name binder (envBindings env)}
+
+{-# ANN addListOfBindings "HLint: ignore Eta reduce" #-}
```

Comment:
Why? `addListOfBindings` would be quite readable applying Eta reduction...

Reply:
Yeah we added these for all places where HLint wanted to Eta reduce. Might be worth reconsidering.